mongo/integration/client_side_encryption_prose_test.go

```diff
@@ -896,6 +897,119 @@ func TestClientSideEncryptionProse(t *testing.T) {
+			[]startedEvent{{"find", "keyvault"}},
+		},
+		{
+			"deadlock case 3", 1, true, false, 2,
+			[]startedEvent{{"find", "db"}, {"find", "keyvault"}},
+			nil,
+		},
+		{
+			"deadlock case 4", 1, true, true, 1,
+			[]startedEvent{{"find", "db"}},
+			[]startedEvent{{"find", "keyvault"}},
+		},
+		{
+			"deadlock case 5", 0, false, false, 1,
+			[]startedEvent{{"listCollections", "db"}, {"listCollections", "keyvault"}, {"find", "keyvault"}, {"insert", "db"}, {"find", "db"}},
+			nil,
+		},
+		{
+			"deadlock case 6", 0, false, true, 1,
+			[]startedEvent{{"listCollections", "db"}, {"insert", "db"}, {"find", "db"}},
+			[]startedEvent{{"find", "keyvault"}},
+		},
+		{
+			"deadlock case 7", 0, true, false, 1,
+			[]startedEvent{{"find", "db"}, {"find", "keyvault"}},
+			nil,
+		},
+		{
+			"deadlock case 8", 0, true, true, 1,
+			[]startedEvent{{"find", "db"}},
+			[]startedEvent{{"find", "keyvault"}},
+		},
+	}
+
+	for _, tc := range testcases {
+		mt.Run(tc.description, func(mt *mtest.T) {
+			var clientEncryptedEvents []startedEvent
+			var clientEncryptedTopologyOpening int
+
+			d := newDeadlockTest(mt)
+			defer d.disconnect(mt)
+
+			kmsProviders := map[string]map[string]interface{}{
+				"local": {"key": localMasterKey},
+			}
+			aeOpts := options.AutoEncryption()
+			aeOpts.SetKeyVaultNamespace("keyvault.datakeys").
+				SetKmsProviders(kmsProviders).
+				SetBypassAutoEncryption(tc.bypassAutoEncryption)
+			if tc.keyVaultClientSet {
+				aeOpts.SetKeyVaultClientOptions(d.clientKeyVaultOpts)
+			}
+
+			ceOpts := options.Client().ApplyURI(mtest.ClusterURI()).
+				SetMonitor(&event.CommandMonitor{
+					Started: func(ctx context.Context, event *event.CommandStartedEvent) {
+						clientEncryptedEvents = append(clientEncryptedEvents, startedEvent{event.CommandName, event.DatabaseName})
+					},
+				}).
+				SetServerMonitor(&event.ServerMonitor{
+					TopologyOpening: func(event *event.TopologyOpeningEvent) {
+						clientEncryptedTopologyOpening++
+					},
+				}).
+				SetMaxPoolSize(tc.maxPoolSize).
+				SetAutoEncryptionOptions(aeOpts)
+
+			clientEncrypted, err := mongo.Connect(mtest.Background, ceOpts)
+			assert.Nil(mt, err, "Connect error: %v", err)
+
+			coll := clientEncrypted.Database("db").Collection("coll")
+			if !tc.bypassAutoEncryption {
+				_, err = coll.InsertOne(mtest.Background, bson.M{"_id": 0, "encrypted": "string0"})
+			} else {
+				unencryptedColl := d.clientTest.Database("db").Collection("coll")
+				_, err = unencryptedColl.InsertOne(mtest.Background, bson.M{"_id": 0, "encrypted": d.ciphertext})
+			}
+			assert.Nil(mt, err, "InsertOne error: %v", err)
+
+			res := coll.FindOne(mtest.Background, bson.M{"_id": 0})
+			assert.Nil(mt, res.Err(), "FindOne error: %v", res.Err())
+
+			raw, err := res.DecodeBytes()
+			assert.Nil(mt, err, "DecodeBytes error: %v", err)
+			expected, _ := bson.Marshal(bson.D{{"_id", 0}, {"encrypted", "string0"}})
```

Comment:
It's generally bad practice to ignore errors, even if you know that the `Marshal` call won't error on this document. I'd recommend using the `bsoncore.NewDocumentBuilder()` here to construct the BSON bytes directly because the document is simple.

Reply:
Agreed, ignoring the error was an unintended mistake. I did not know about the `DocumentBuilder`. Done.